### DISCUSSION TITLE
Printer tests rework

### DISCUSSION THREAD

**Discussion by belono** [python-escpos/python-escpos#574] [Show and tell]
This is an attempt to rework the tests battery for the File printer that I think it could be extended (in essence) to the rest of the printer connectors.
I've tried in my local machine and seems to work, all this tests are passing.

```python
# test_printer_file.py

import logging

import pytest


def test_device_not_initialized(file):
    """
    GIVEN a file printer object
    WHEN it is not initialized
    THEN check the device property is False
    """
    assert file._device is False


def test_open_raise_exception(file, devicenotfounderror):
    """
    GIVEN a file printer object
    WHEN open() is set to raise a DeviceNotFoundError on error
    THEN check the exception is raised
    """
    file.devfile = "fake/device"

    with pytest.raises(devicenotfounderror):
        file.open(raise_not_found=True)


def test_open_not_raise_exception(file, caplog):
    """
    GIVEN a file printer object
    WHEN open() is set to not raise on error but simply cancel
    THEN check the error is logged and open() canceled
    """
    file.devfile = "fake/device"
    with caplog.at_level(logging.ERROR):
        file.open(raise_not_found=False)

    assert "not found" in caplog.text
    assert file.device is None


def test_open(file, caplog):
    """
    GIVEN a file printer object
    WHEN a valid connection to a device is opened
    THEN check the success is logged and the device property is set
    """
    file.devfile = "/dev/null"
    with caplog.at_level(logging.INFO):
        file.open()

    assert "enabled" in caplog.text
    assert file.device


def test_flush(file, mocker):
    """
    GIVEN a file printer object
    WHEN auto_flush is disabled and flush() issued manually
    THEN check the flush method is called only one time.
    """
    spy = mocker.spy(file, "flush")
    file.devfile = "/dev/null"
    file.auto_flush = False
    file.open()
    file.textln("python-escpos")
    file.flush()

    assert spy.call_count == 1


def test_auto_flush_on_command(file, mocker):
    """
    GIVEN a file printer object
    WHEN auto_flush is enabled and flush() not issued manually
    THEN check the flush method is called automatically
    """
    spy = mocker.spy(file, "flush")
    file.devfile = "/dev/null"
    file.auto_flush = True
    file.open()
    file.textln("python-escpos")
    file.textln("test")

    assert spy.call_count > 1


def test_auto_flush_on_close(file, mocker, caplog, capsys):
    """
    GIVEN a file printer object
    WHEN auto_flush is disabled and flush() not issued manually
    THEN check the flush method is called automatically on close
    """
    spy = mocker.spy(file, "flush")
    file.devfile = "/dev/null"
    file.auto_flush = False
    file.open()
    file.textln("python-escpos")
    file.close()

    assert spy.call_count == 1


def test_close(file, caplog):
    """
    GIVEN a file printer object
    WHEN a connection is opened and closed
    THEN check the closing is logged and the device property is False
    """
    file.devfile = "/dev/null"
    file.open()
    with caplog.at_level(logging.INFO):
        file.close()

    assert "Closing" in caplog.text
    assert file._device is False

```
Downloadable version: [test_printer_file.py.zip](https://github.com/python-escpos/python-escpos/files/12729426/test_printer_file.py.zip)

The `file` and `devicenotfounderror` fixtures would be found in conftest.py with other printer connector fixtures:
```python
# conftest.py

from escpos.exceptions import DeviceNotFoundError
from escpos.printer import LP, CupsPrinter, Dummy, File, Network, Serial, Usb, Win32Raw


@pytest.fixture
def driver():
    return Dummy()


@pytest.fixture
def file():
    return File()


@pytest.fixture
def devicenotfounderror():
    return DeviceNotFoundError
```

As the `File` printer points to a file object it is easy to make it write to `/dev/null` or to a `temp_file` to simulate a printer, but obviously it is not as easy to do on the other printer connectors.

What do you think about this approach?

**Comment by belono** [comment]
I've been curious about how 'pyserial' and 'pyusb' perform their tests and I've found some interesting stuff:

The `pyserial` library emulates a serial connection using the `pty.openpty()` and the `os.ttyname()` functions from the python standard library to get a fake file descriptor to read and write to. Look at [pyserial test_pty](https://github.com/pyserial/pyserial/blob/7aeea35429d15f3eefed10bbb659674638903e3a/test/test_pty.py#L24).

The `pyusb` project seems to take a different strategy and they create a fake usb device descriptor object attached to a custom back-end, then they can successfully `usb.core.find(backend=custom_backend, idVendor=fake_usb_device_descriptor_idVendor)`. Look at [pyusb test_find](https://github.com/pyusb/pyusb/blob/5b191b7af2025528f2a5395c0077f599a9c3d670/tests/test_find.py#L39).

We could inject or mock one of this fake devices into a `printer.Serial()` object and a `printer.Usb()` object respectively during testing.

**Reply by patkan** [reply]
Sounds good. On the other hand wouldn't it be enough for our purpose to mock the complete pyusb or pyserial Library and check if we call the right functions? We do not really care (for a unit test), whether the underlying library actually sends the correct information on the Bus.

Ideally we would have some additional integration test script that could be executed on a printer and checked manually, but I just have one printer and the fact that this could not be automated is also not good.

**Reply by belono** [reply]
> On the other hand wouldn't it be enough for our purpose to mock the complete pyusb or pyserial Library and check if we call the right functions? We do not really care (for a unit test), whether the underlying library actually sends the correct information on the Bus.

You are right, it is overkill.
Mock the device and test the printer's functions seems to be the easiest solution, plus it is extensible to the other printers' connectors unlike the other approach.

**Comment by belono** [comment]
Following your suggestion I reworked the previous tests to make use of the mocker fixture and also adapted them to the Serial printer:

Note that I don't know how to force the mock device to raise a valid exception that triggers the DeviceNotFoundError, so I don't mock the device on those tests.

Maybe some tests are not necessary.
I'm open to any suggestion.

```python
def test_device_not_initialized(serial):
    """
    GIVEN a serial printer object
    WHEN it is not initialized
    THEN check the device property is False
    """
    assert serial._device is False


def test_open_raise_exception(serial, devicenotfounderror):
    """
    GIVEN a serial printer object
    WHEN open() is set to raise a DeviceNotFoundError on error
    THEN check the exception is raised
    """
    serial.devfile = "fake/device"

    with pytest.raises(devicenotfounderror):
        serial.open(raise_not_found=True)


def test_open_not_raise_exception(serial, caplog):
    """
    GIVEN a serial printer object
    WHEN open() is set to not raise on error but simply cancel
    THEN check the error is logged and open() canceled
    """
    serial.devfile = "fake/device"
    with caplog.at_level(logging.ERROR):
        serial.open(raise_not_found=False)

    assert "not found" in caplog.text
    assert serial.device is None


def test_open(serial, caplog, mocker):
    """
    GIVEN a serial printer object and a mocked pyserial device
    WHEN a valid connection to a device is opened
    THEN check the success is logged and the device property is set
    """
    mocker.patch("serial.Serial")
    with caplog.at_level(logging.INFO):
        serial.open()

    assert "enabled" in caplog.text
    assert serial.device


def test_close(serial, caplog, mocker):
    """
    GIVEN a serial printer object and a mocked pyserial device
    WHEN a connection is opened and closed
    THEN check the closing is logged and the device property is False
    """
    mocker.patch("serial.Serial")
    serial.open()
    with caplog.at_level(logging.INFO):
        serial.close()

    assert "Closing" in caplog.text
    assert serial._device is False

```
All this tests are actually passing OK.

**Comment by belono** [comment]
If no objections, I'm going to open a new PR adding the reworked printer tests. 
We can continue this discussion during the review of the changes. ;-)

**Reply by patkan** [reply]
Yes, please do that
